compiler/rustc_errors/src/diagnostic.rs

```diff
@@ -630,18 +630,27 @@ impl Diagnostic {
         style: SuggestionStyle,
     ) -> &mut Self {
         assert!(!suggestion.is_empty());
-        debug_assert!(
-            !(suggestion.iter().any(|(sp, text)| sp.is_empty() && text.is_empty())),
-            "Span must not be empty and have no suggestion"
-        );
+
+        let parts = suggestion
+            .into_iter()
+            .map(|(span, snippet)| SubstitutionPart { snippet, span })
+            .collect::<Vec<_>>();
+
+        for (index, part_outer) in parts.iter().enumerate() {
+            debug_assert!(
+                !part_outer.span.is_empty() || !part_outer.snippet.is_empty(),
+                "Span must not be empty and have no suggestion",
+            );
+            for part_inner in parts.iter().skip(index + 1) {
+                debug_assert!(
+                    !part_outer.span.overlaps(part_inner.span),
+                    "suggestion must not have overlapping parts:\n{part_outer:?}\n{part_inner:?}",
+                );
+            }
+        }
```

Comment:
Does it make sense to sort parts by the `span`? this way you only need to check the two adjacent ones. But, we shouldn't do this just for the debug assertions maybe...

Either way, I think it would be better to enclose this in `if cfg!(debug_assertions) {}`, so llvm has an easier time removing the useless loops.

Also, can we somehow test this assertion?

Reply:
I think it is perfectly safe to always unconditionally sort here.